### DISCUSSION TITLE
Can't find include file "emqx_coap.hrl"

### DISCUSSION THREAD

**Discussion by sixty-four-bit** [emqx/emqx#4660] [Q&A]
HI , I am using VS Code to compile emqx.  But it fails to find the header files for all app if the header files are in include folder. 

===> Compiling apps/emqx_coap/src/emqx_coap_timer.erl failed
apps/emqx_coap/src/emqx_coap_timer.erl:19: can't find include file "emqx_coap.hrl"

Tried on both VSCode on Mac and VSCode on Windows 10.

launch.json
{
    // Use IntelliSense to learn about possible attributes.
    // Hover to view descriptions of existing attributes.
    // For more information, visit: https://go.microsoft.com/fwlink/?linkid=830387
    "version": "0.2.0",
    "configurations": [
        {
            "name": "Launch erlang",
            "type": "erlang",
            "request": "launch",
            "cwd": "${workspaceRoot}",
            "preLaunchTask": "rebar3 compile"
        }
    ]
}

tasks.json
{
    // See https://go.microsoft.com/fwlink/?LinkId=733558
    // for the documentation about the tasks.json format
    "version": "2.0.0",
    "tasks": [
      {
        "label": "rebar3 compile",
        "type": "shell",
        "command": "rebar3 compile",
        "group": {
          "kind": "build",
          "isDefault": true
        },
        "problemMatcher": "$erlang"
      }
    ]
  }

Any help appreciated

**Comment by zmstone** [comment]
EMQ X project only supports makefile compile.
I’d recommend to configure vscode to run `make` instead of rebar3 directly.

Or maybe (I have no experience, so only guessing here) you can try running scripts/ensure-rebar3.sh make sure it’s downloaded in project root, then try to build again in vs code
Edit: we use a rebar3 fork for emqx

**Reply by sixty-four-bit** [reply]
Hi zmstone.

Thanks for your reply.  I am new to Emqx and Erlang. 

In that case  how can I debug emqx.

**Reply by zmstone** [reply]
I am not sure how vscode builds an Erlang project, I'd assume it uses rebar3 (the most popular builder for Erlang).
If this is the case, `./scripts/ensure-rebar3.sh` should download the builder to the project root, then vscode should be able to build the project.

Since you are trying to debug emqx,
could you be more specific on what you want to debug?
most of the beginners try to add debug logs in the code and inspect the printouts.

There is no breakpoint/stepping through kind of debugging in Erlang.
but the Erlang shell REPL is quite powerful once you know some Erlang basics.
the book for beginner: https://learnyousomeerlang.com/

**Comment by zmstone** [comment]
without vscode.
`make` to build a release
`make run` to run a node.
